### PR TITLE
Allow adding an optional description for app repositories #759

### PR DESCRIPTION
**Pull request by gfichtenholt**
Allow adding an optional description for app repositories

### Applicable issues

  - fixes #

### CODE REVIEW

File: cmd/apprepository-controller/hack/update-codegen.sh
Comment (absoludity) [inline cmd/apprepository-controller/hack/update-codegen.sh:35]: Sorry, I'd not realised on the slack conv. that you were explicitly asking after modifying types.go - my answer was when does it need to be run generally (which is when we update the client-go library). If you update `types.go` and add any structured types, it'd be required as the generated code updates the deep-copy implementation (for copying structs with pointers to other structs, rather than copying the ref), which is not necessary for a string field (not 100%, but your code is working).

```suggestion
# What are the situations when one needs to run update-codegen.sh manually after modifying 
# types.go in apprepository-controller?
# Generally if the generated client will change. This can happen for multiple reasons. If you
# change to the `AppRepository.Spec` struct the updated struct is imported by the client (so
# regeneration not needed necessarily) but if your change has a complex type (structs and 
# pointers to structs) the generated client will need to be updated as it includes a deep copy
# function. Another cause of your generated client changing is if you update the version of the
# client.go library that your project uses.
# Note that the generated client is generally stored in the repo so that other (external) projects
# can import your client directly. You can see examples of generated clients in the sample-controller
# for the the K8s org at:
#  https://github.com/kubernetes/sample-controller/blob/master/pkg/generated/clientset/versioned/typed/samplecontroller/v1alpha1/samplecontroller_client.go
# or any other controllers in the k8s project.
```

Hope that's clearer.

**Review by absoludity — COMMENTED** [review]
Thanks Greg! Great first PR. I've a couple of uncertainties:
* Not sure why you're updating the signature of the validation function (I can see that it could have been understood to be validating the name and description fields, but it doesn't, it validates that the actual URL can be accessed with the provided details).
* Similarly, not sure why you're including the description in the args to the sync job (which doesn't use it at all).

Otherwise, looks great. Glad you didn't have to add test infra but could just add to the existing suites (albeit some being a little opaque, as commented).

+1 once we're agreed about those two places where description is being added unnecessarily (unless I'm mistaken).

File: cmd/asset-syncer/main.go
Comment (gfichtenholt) [reply cmd/asset-syncer/main.go:63]: same answer as above

File: cmd/apprepository-controller/controller.go
Comment (gfichtenholt) [reply cmd/apprepository-controller/controller.go:623]: me neither. I was just trying to be consistent with the code that was already there. why do any of the other fields, like --tls-insecure-skip-verify or --filter-rules were already there? not sure. probably by mistake? Anyway, I can remove it from my file set

File: dashboard/src/actions/repos.test.tsx
Comment (gfichtenholt) [reply dashboard/src/actions/repos.test.tsx:1053]: I see. I figured some day we might want to do some validation of description on the back end, so I added. If you feel it would be cleaner not to do it today, I will remove description from validate() calls.

File: dashboard/src/shared/AppRepository.ts
Comment (gfichtenholt) [reply dashboard/src/shared/AppRepository.ts:118]: same as above. I am fine to remove it

File: dashboard/src/actions/repos.ts
Comment (gfichtenholt) [reply dashboard/src/actions/repos.ts:329]: same as above

File: dashboard/src/components/Config/AppRepoList/AppRepoForm.tsx
Comment (gfichtenholt) [reply dashboard/src/components/Config/AppRepoList/AppRepoForm.tsx:223]: ok

File: dashboard/src/components/Config/AppRepoList/AppRepoList.test.tsx
Comment (gfichtenholt) [reply dashboard/src/components/Config/AppRepoList/AppRepoList.test.tsx:218]: ok

File: dashboard/src/components/Config/AppRepoList/AppRepoList.tsx
Comment (gfichtenholt) [reply dashboard/src/components/Config/AppRepoList/AppRepoList.tsx:248]: ok

File: dashboard/src/shared/AppRepository.ts
Comment (gfichtenholt) [reply dashboard/src/shared/AppRepository.ts:131]: already commented above

File: pkg/kube/kube_handler_test.go
Comment (gfichtenholt) [reply pkg/kube/kube_handler_test.go:320]: ok

File: cmd/apprepository-controller/hack/update-codegen.sh
Comment (gfichtenholt) [reply cmd/apprepository-controller/hack/update-codegen.sh:35]: ok

File: cmd/apprepository-controller/controller_test.go
Comment (absoludity) [inline cmd/apprepository-controller/controller_test.go:1256]: I think you missed removing this test? It's testing that the description is included in the args to the sync job (and is causing CI to fail for your PR).

File: cmd/apprepository-controller/controller.go
Comment (absoludity) [reply cmd/apprepository-controller/controller.go]: The other args are there because they're used to sync chart metadata from the repository (eg. the `--tls-insecure-skip-verify` is an option if the tls cert for the repo should not be verified when contacting the repo, the `--filter-rules` are used to query an OCI repo, etc.). We just don't need the description when syncing the metadata :)

File: dashboard/src/actions/repos.test.tsx
Comment (absoludity) [reply dashboard/src/actions/repos.test.tsx]: Great, thanks. Yes, I'd add it as an arg to `validateRepo` only when we're going to actually use it.

File: cmd/apprepository-controller/controller_test.go
Comment (gfichtenholt) [reply cmd/apprepository-controller/controller_test.go]: yes, indeed. fixed now

File: cmd/apprepository-controller/controller.go
Comment (gfichtenholt) [reply cmd/apprepository-controller/controller.go]: ok, got it

File: dashboard/src/actions/repos.test.tsx
Comment (gfichtenholt) [reply dashboard/src/actions/repos.test.tsx]: ok

**Review by absoludity — APPROVED** [review]
Excellent, thanks for the updates Greg. +1, but as you noted, CI is failing for unrelated reasons (there was a new release of the bitnami kubeapps chart in the past hours, as part of #2704 , for which we'll need to update our CI script). Once that change lands we can retrigger the CI job for your PR here which should pass.

**Review by antgamdia — APPROVED** [review]
+1 Thank you for the changes and for the thorough review!
Indeed, I wasn't fully aware of the `update-codegen.sh`, so, thanks for the clarification in the file.

As for me, I've been fixing the CI failures (derived from drastically changing the way in which we synchronize our chart with the bitnami one), so now it should work again.

Please, @gfichtenholt , merge the latest changes in master into your branch so that you get the fix and the PR test passes.

**Comment by gfichtenholt** [comment]
I don't have any more pending changes, all changes were committed yesterday. I can commit a no-op change to trigger the pipeline if you like

**Comment by antgamdia** [comment]
> I don't have any more pending changes, all changes were committed yesterday. I can commit a no-op change to trigger the pipeline if you like

No, I just meant that you merge the upstream changes into your local branch (in your fork). Let me explain it:

Assuming you are in your fork `gfichtenholt/kubeapps`, you might have set two different remotes. You can check it by `git remote -v `

```console
git remote -v

origin  https://github.com/antgamdia/kubeapps.git (fetch)
origin  https://github.com/antgamdia/kubeapps.git (push)

upstream        git@github.com:kubeapps/kubeapps.git (fetch)
upstream        git@github.com:kubeapps/kubeapps.git (push)
```

As you can see, I have the `origin` remote pointing to my fork, but also, I have the `upstream` remote pointing to the `kubeapps/kubeapps` main repository.

If you don't have, add it by simply executing:

```bash
git remote add upstream git@github.com:kubeapps/kubeapps.git 
```

Once you have it configured, just pull the latest changes in (upstream) master by:

```
 git pull upstream  master
```

Then, push the changes to your local remote branch (`git push`) and that's all! You will have retrieved the latest changes in the main repo and merged them into your branch. 

Note that this is not something common, but in this case, you need to get a change in the `e2e-test.sh` and `chart_sync_utils.sh` so that your PR doesn't fail.

**Comment by gfichtenholt** [comment]
Antonio, done as you instructed. Thank you

**Comment by antgamdia** [comment]
(@gfichtenholt, feel free to merge this PR whenever you want by clicking on the "squash and merge" button)
Also, if you edit the PR message, you can add:

```markdown
- fixes #759 
```

so that the associated issue (#759) will automatically get closed once you merge it.

**Comment by gfichtenholt** [comment]
Antonio, I would if I could. I don't have a 'squash and merge' button. My guess is I don't have write access to the repo.

**Comment by antgamdia** [comment]
@gfichtenholt , would you mind trying again now? I've been configuring the repo permissions

**Comment by gfichtenholt** [comment]
it worked. Muchas gracias!

**Pull request merged**
